File: sdks/python/src/echo_agent_sdk/skill_content_values.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SkillResourceEntry:
    kind: str
    relative_path: str


@dataclass(frozen=True, slots=True)
class SkillContent:
    name: str
    skill_dir: str
    instructions: str
    allowed_tools: tuple[str, ...] = ()
    resources: tuple[SkillResourceEntry, ...] = ()

    def to_prompt_block(self) -> str:
        block = (
            f'<skill_content name="{self.name}">\n{self.instructions.strip()}\n\n'
            f"Skill directory: {self.skill_dir}\n"
            "Relative paths in this skill are relative to the skill directory."
        )
        if self.allowed_tools:
            block += (
                "\n\n<allowed_tools>\nThis skill declares the following preferred/allowed tools: "
                f"{', '.join(self.allowed_tools)}\nRuntime enforcement currently applies to the built-in skill tools "
                "such as read_skill_resource and run_skill_script.\n</allowed_tools>"
            )
        if self.resources:
            block += "\n\n<skill_resources>"
            for resource in self.resources:
                block += (
                    f'\n  <file kind="{resource.kind}">{resource.relative_path}</file>'
                )
            block += "\n</skill_resources>"
        return f"{block}\n</skill_content>"

File: sdks/typescript/src/skill_content_values.ts
/** Activated Skill content projection without loading or executing resources. */
export interface SkillResourceEntry {
  readonly kind: string;
  readonly relativePath: string;
}

export class SkillContent {
  public constructor(
    public readonly name: string,
    public readonly skillDir: string,
    public readonly instructions: string,
    public readonly allowedTools: readonly string[] = [],
    public readonly resources: readonly SkillResourceEntry[] = [],
  ) {}

  public toPromptBlock(): string {
    let block = `<skill_content name="${this.name}">\n${this.instructions.trim()}\n\nSkill directory: ${this.skillDir}\nRelative paths in this skill are relative to the skill directory.`;
    if (this.allowedTools.length > 0) {
      block += `\n\n<allowed_tools>\nThis skill declares the following preferred/allowed tools: ${this.allowedTools.join(", ")}\nRuntime enforcement currently applies to the built-in skill tools such as read_skill_resource and run_skill_script.\n</allowed_tools>`;
    }
    if (this.resources.length > 0) {
      block += "\n\n<skill_resources>";
      for (const resource of this.resources) block += `\n  <file kind="${resource.kind}">${resource.relativePath}</file>`;
      block += "\n</skill_resources>";
    }
    return `${block}\n</skill_content>`;
  }
}
